# Proyecto RappiPlus: de datos a decisiones de negocio

**Introducción**


El objetivo de este proyecto es evaluar el desempeño del servicio **RappiPlus** para apoyar **decisiones de negocio basadas en datos**.

Se trabajan con múltiples datasets del negocio:

- **rappiplus_orders_raw.csv** → información de pedidos, precios, descuentos y revenue  
- **rappiplus_catalog.csv** → costos de productos, categorías y proveedores  
- **rappiplus_marketing_spend.csv** → inversión en marketing por canal y país  
- **events / users / user_activity (SQL)** → comportamiento del usuario dentro de la plataforma  
- **experiment_checkout_ui.csv** → resultados de un experimento A/B en el checkout  

El análisis sigue una lógica clara y progresiva:

1. 🔍 Evaluar si podemos confiar en los datos (calidad de datos en Python) 

2. 💰 Analizar si el negocio es rentable (revenue, costos y profit)  

3. 🛒 Entender dónde se pierden los usuarios (funnel de conversión)  

4. 🔁 Evaluar si los usuarios regresan (retención por cohortes)  

5. 🧪 Validar si los cambios generan impacto (test estadístico)  

6. 📊 Comunicar los resultados (dashboard en BI)  

A lo largo del proyecto, se transforman datos en insights para responder preguntas clave del negocio y proponer **recomendaciones accionables**.

---

## 🔹 Paso 1: Cargar y validar la calidad de los datos

---

### 1.1 Carga de datos y vista rápida

**🎯 Objetivo:** Familiarizarte con la estructura de los datasets del negocio antes de analizarlos.

**Instrucciones:**

- Importa las librerías necesarias
- Carga los archivos:
  - `rappiplus_orders_raw.csv`
  - `rappiplus_catalog.csv`
  - `rappiplus_marketing_spend.csv`
- Guarda los DataFrames en:
  - `orders`, `catalog`, `marketing`
- Explora cada dataset.

---

In [1]:
# importar librerías
import pandas  as pd               #LECTURA, MANEJO Y TRANSFORMACIÓN DE DATAFRAMES.
import numpy   as np               #FUNCIONES MATEMÁTICAS RELEVANTES.
import matplotlib.pyplot as plt    #GRÁFICOS.
import seaborn as sns              #GRÁFICOS PRETTY.
import unicodedata

In [2]:
# cargar archivos
orders = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_orders_raw.csv')
catalog = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_catalog.csv')
marketing = pd.read_csv('https://practicum-content.s3.amazonaws.com/datasets/rappiplus_marketing_spend.csv')

In [3]:
# explorar datasets

In [4]:
orders.head(5)

,id_pedido,id_usuario,fecha_hora_pedido,pais,dispositivo,fuente_referencia,nombre_producto,categoria_producto,cantidad,precio_unitario,monto_descuento,monto_total
0,order_0,user_6993,2025-05-22,Argentina,desktop,organic,Jacket-Winter-M,Moda,2.0,332.69,0.0,665.37
1,order_1,user_1329,2025-06-15,Mexico,desktop,paid_search,Tablet-Standard-64GB,Electronica,1.0,176.86,5.0,171.86
2,order_2,user_3194,2025-05-02,Argentina,desktop,social,Blender-XL-Red,Hogar,2.0,102.99,10.0,195.99
3,order_3,user_4510,2025-06-09,Colombia,mobile,social,Tablet-Standard-64GB,Electronica,1.0,257.87,15.0,242.87
4,order_4,user_5044,2025-03-30,Argentina,desktop,paid_search,Blender-XL-Red,Hogar,1.0,336.28,0.0,336.28


In [5]:
catalog.head(5)

,nombre_producto,categoria_producto,costo_unitario,proveedor
0,Laptop-Gaming-16GB,Electrónica,280.68,"Fuller, Pena and Myers"
1,Phone-Pro-128GB,Electrónica,10.12,King Ltd
2,Tablet-Standard-64GB,Electrónica,25.21,Bowers LLC
3,Blender-XL-Red,Hogar,176.64,Long-Reid
4,Vacuum-Pro-Black,Hogar,16.60,"Rivera, Carr and Finley"


In [6]:
marketing.head(5)

,fecha,pais,id_campaña,canal,gasto
0,2025-01-01,Mexico,organic_Mexico,organic,2446.25
1,2025-01-01,Mexico,paid_search_Mexico,paid_search,2704.34
2,2025-01-01,Mexico,social_Mexico,social,2045.01
3,2025-01-01,Colombia,organic_Colombia,organic,2597.21
4,2025-01-01,Colombia,paid_search_Colombia,paid_search,1771.40


In [7]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25100 entries, 0 to 25099
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   id_pedido           25100 non-null  object 
 1   id_usuario          25100 non-null  object 
 2   fecha_hora_pedido   25100 non-null  object 
 3   pais                24800 non-null  object 
 4   dispositivo         25080 non-null  object 
 5   fuente_referencia   25070 non-null  object 
 6   nombre_producto     25070 non-null  object 
 7   categoria_producto  25020 non-null  object 
 8   cantidad            25050 non-null  float64
 9   precio_unitario     25050 non-null  float64
 10  monto_descuento     25050 non-null  float64
 11  monto_total         25100 non-null  float64
dtypes: float64(4), object(8)
memory usage: 2.3+ MB


### Dataset: rappiplus_orders_raw
25,100 filas. Cada fila = un pedido en la plataforma RappiPlus.
Contiene info de usuario, fecha, país, producto, cantidad, precios y descuentos.

In [8]:
catalog.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7 entries, 0 to 6
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   nombre_producto     7 non-null      object 
 1   categoria_producto  7 non-null      object 
 2   costo_unitario      7 non-null      float64
 3   proveedor           7 non-null      object 
dtypes: float64(1), object(3)
memory usage: 352.0+ bytes


### Dataset: rappiplus_catalog.csv
7 filas. Cada fila representa un producto disponible en la plataforma RappiPlus.
Contiene info de nombre_producto, categoria_producto, costo_unitario, proveedor.

In [9]:
marketing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1620 entries, 0 to 1619
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   fecha       1620 non-null   object 
 1   pais        1620 non-null   object 
 2   id_campaña  1620 non-null   object 
 3   canal       1519 non-null   object 
 4   gasto       1620 non-null   float64
dtypes: float64(1), object(4)
memory usage: 63.4+ KB


### Dataset: rappiplus_marketing_spend
1620 filas. Cada fila representa una inversión en marketing realizada en un país y canal específico en la plataforma RappiPlus.
Contiene info de nombre_producto, categoria_producto, costo_unitario, proveedor.

---

### Revisión y calidad de datos

**🎯 Objetivo:** Detectar y corregir problemas en los datos que puedan afectar el análisis de revenue, costos y rentabilidad.

Se revisan los 3 datasets
- Validar y convertir fechas al formato correcto  
- Revisar variables numéricas (sin negativos o ceros inválidos)  
- Verifica
r consistencia de montos  
- Eliminar duplicados  
- Revisar variables categóricas 


---

### Validar y convertir fechas

In [10]:
# Convertir fecha_hora_pedido de texto a datetime real
orders['fecha_hora_pedido'] = pd.to_datetime(orders['fecha_hora_pedido'], errors='coerce')


In [11]:
# Verificar si quedó algún nulo tras la conversión (fechas mal escritas)
print(orders['fecha_hora_pedido'].isnull().sum())

0


In [12]:

# Revisar el rango de fechas para detectar algo fuera de lo esperado
print(orders['fecha_hora_pedido'].min(), '→', orders['fecha_hora_pedido'].max())


2025-01-01 00:00:00 → 2025-06-30 00:00:00


### Revisar variables numéricas (negativos / ceros inválidos)

In [13]:
columnas_numericas = ['cantidad', 'precio_unitario', 'monto_descuento', 'monto_total']

# Ver estadísticas generales: mínimo, máximo, promedio
print(orders[columnas_numericas].describe())

# Contar cuántas filas tienen cantidad negativa
print('Cantidad negativa:', (orders['cantidad'] < 0).sum())

# Contar cuántas filas tienen precio_unitario en cero o negativo (inválido, un producto no puede costar $0 o menos)
print('Precio inválido:', (orders['precio_unitario'] <= 0).sum())

# Ver si el descuento es mayor al valor del producto (inconsistente)
print('Descuento mayor al precio:', (orders['monto_descuento'] > orders['precio_unitario']).sum())


# Bandera de devoluciones (cantidad negativa) — se crea AQUÍ, junto con el resto de banderas
orders['es_devolucion'] = orders['cantidad'] < 0
print(orders['es_devolucion'].sum(), 'pedidos marcados como devolución')


           cantidad  precio_unitario  monto_descuento   monto_total
count  25050.000000     25050.000000     25050.000000  2.510000e+04
mean       7.092735       259.305549         4.500798  2.072680e+03
std      296.277003       138.726461         5.223010  9.894995e+04
min       -2.000000        20.030000         0.000000 -4.926500e+02
25%        1.000000       138.377500         0.000000  1.805075e+02
50%        2.000000       258.715000         0.000000  3.417500e+02
75%        2.000000       380.332500        10.000000  5.185800e+02
max    20000.000000       499.960000        15.000000  8.840200e+06
Cantidad negativa: 4
Precio inválido: 0
Descuento mayor al precio: 0
4 pedidos marcados como devolución


### Verificar consistencia de montos

In [14]:

# Fórmula corregida desde el inicio (ya sabemos que las devoluciones invierten el signo del descuento)
orders['monto_calculado'] = (orders['precio_unitario'] * orders['cantidad']) - (orders['monto_descuento'] * np.sign(orders['cantidad']))
orders['diferencia'] = (orders['monto_total'] - orders['monto_calculado']).abs()
orders['monto_inconsistente'] = orders['diferencia'] > 0.02

print(orders['monto_inconsistente'].sum(), 'filas realmente inconsistentes')

0 filas realmente inconsistentes


### Eliminar duplicados

In [15]:

for nombre, df in [('orders', orders), ('catalog', catalog), ('marketing', marketing)]:
    exactos = df.duplicated().sum()
    print(f'{nombre}: {exactos} filas exactamente duplicadas')

duplicados_id = orders['id_pedido'].duplicated().sum()
print(f'id_pedido repetidos: {duplicados_id}')

orders = orders.drop_duplicates()
catalog = catalog.drop_duplicates()
marketing = marketing.drop_duplicates()

orders: 100 filas exactamente duplicadas
catalog: 0 filas exactamente duplicadas
marketing: 0 filas exactamente duplicadas
id_pedido repetidos: 100


###  Revisar variables categóricas

In [16]:
columnas_categoricas = ['pais', 'dispositivo', 'fuente_referencia', 'categoria_producto']

for col in columnas_categoricas:
    print(f'\n--- {col} ---')
    print(orders[col].unique())  # todos los valores distintos que existen


--- pais ---
['Argentina' 'Mexico' 'Colombia' 'mexico' 'colombia' 'argentina' nan]

--- dispositivo ---
['desktop' 'mobile' nan]

--- fuente_referencia ---
['organic' 'paid_search' 'social' nan]

--- categoria_producto ---
['Moda' 'Electronica' 'Hogar' nan]


In [17]:
# Comparar categorías entre orders y catalog
print('En orders:', sorted(orders['categoria_producto'].dropna().unique()))
print('En catalog:', sorted(catalog['categoria_producto'].unique()))

En orders: ['Electronica', 'Hogar', 'Moda']
En catalog: ['Electrónica', 'Hogar', 'Moda']


 Estandarizar pais

In [18]:
orders['pais'] = orders['pais'].str.strip().str.title()

# Verificar que ahora solo existan 3 valores + los nulos
print(orders['pais'].unique())

['Argentina' 'Mexico' 'Colombia' nan]


-**Outliers de cantidad / monto_total — recomendación**

In [19]:
print(orders.sort_values('cantidad', ascending=False).head(10)[['id_pedido','cantidad','precio_unitario','monto_total']])
orders['outlier_cantidad'] = orders['cantidad'] > 10
print(orders['outlier_cantidad'].sum(), 'pedidos marcados como outlier de cantidad')

       id_pedido  cantidad  precio_unitario  monto_total
3668  order_3668   20000.0           348.31    6966200.0
3656  order_3656   20000.0           297.66    5953200.0
3726  order_3726   20000.0           290.85    5817000.0
3722  order_3722   20000.0           442.01    8840200.0
3521  order_3521   10000.0            43.14     431400.0
3748  order_3748   10000.0           336.93    3369300.0
3586  order_3586   10000.0           490.35    4903500.0
3643  order_3643   10000.0           238.15    2381500.0
3689  order_3689   10000.0            87.69     876900.0
3522  order_3522   10000.0           280.55    2805500.0
10 pedidos marcados como outlier de cantidad


**unificar tildes en catalogo**

In [20]:


def quitar_tildes(texto):
    return ''.join(c for c in unicodedata.normalize('NFKD', texto) if not unicodedata.combining(c))


catalog['categoria_producto'] = catalog['categoria_producto'].apply(quitar_tildes)

# Confirmar que ahora coincide con orders

print('En orders:', sorted(orders['categoria_producto'].dropna().unique()))
print('En catalog:', sorted(catalog['categoria_producto'].unique()))


En orders: ['Electronica', 'Hogar', 'Moda']
En catalog: ['Electronica', 'Hogar', 'Moda']


### Reemplazar la columna canal original con la versión corregida ###

### Recuperar categorías faltantes

In [21]:
mapa_categorias = catalog.set_index('nombre_producto')['categoria_producto'].to_dict()

orders['categoria_producto'] = orders['categoria_producto'].fillna(
    orders['nombre_producto'].map(mapa_categorias)
)
print(orders['categoria_producto'].isnull().sum(), 'nulos restantes en categoria_producto')

30 nulos restantes en categoria_producto


**Nulos restantes — recomendación**

- Rellenar con "Desconocido"

In [22]:
# Bandera de datos incompletos (definida ANTES de rellenar nada, para que mida el estado real)
orders['dato_incompleto'] = orders[['nombre_producto','cantidad','precio_unitario']].isnull().any(axis=1)

# Rellenar categóricas de bajo impacto financiero
for col in ['dispositivo', 'fuente_referencia', 'pais']:
    orders[col] = orders[col].fillna('Desconocido')

print(orders['dato_incompleto'].sum(), 'filas con datos incompletos, excluidas del cálculo de revenue')

80 filas con datos incompletos, excluidas del cálculo de revenue


## remplazo # 

In [23]:
print(marketing['id_campaña'].head())
print(marketing.columns.tolist())

0          organic_Mexico
1      paid_search_Mexico
2           social_Mexico
3        organic_Colombia
4    paid_search_Colombia
Name: id_campaña, dtype: object
['fecha', 'pais', 'id_campaña', 'canal', 'gasto']


In [24]:
marketing['canal_corregido'] = marketing['id_campaña'].str.rsplit('_', n=1).str[0]
print(marketing['canal_corregido'].unique())

['organic' 'paid_search' 'social']


In [25]:
print(marketing.columns.tolist())

['fecha', 'pais', 'id_campaña', 'canal', 'gasto', 'canal_corregido']


In [26]:
# --- Paso 2: recrear canal_corregido a partir de id_campaña ---
marketing['canal_corregido'] = marketing['id_campaña'].str.rsplit('_', n=1).str[0]
# --- Paso 3: confirmar que quedó bien, ANTES de sobrescribir canal ---
print(marketing['canal_corregido'].unique())

['organic' 'paid_search' 'social']


In [27]:
# --- Paso 4: reemplazar canal (que tenía el bug) con la versión corregida ---
marketing['canal'] = marketing['canal_corregido']
marketing = marketing.drop(columns=['canal_corregido'])

In [28]:
print(marketing['canal'].unique())
print(marketing.isnull().sum())

['organic' 'paid_search' 'social']
fecha         0
pais          0
id_campaña    0
canal         0
gasto         0
dtype: int64


In [29]:
marketing.to_csv('marketing_clean.csv', index=False)

# Verificación de que se guardó bien
prueba_marketing = pd.read_csv('marketing_clean.csv')
print(prueba_marketing['canal'].unique())

['organic' 'paid_search' 'social']


In [30]:

print(orders.shape)          # confirmar el número de filas/columnas final
print(orders.columns.tolist())  # confirmar que existen las columnas bandera (es_devolucion, outlier_cantidad, etc.)
print(orders.isnull().sum())    # última revisión de nulos restantes

print(catalog.shape)
print(marketing.shape)
print(marketing.isnull().sum())


(25000, 18)
['id_pedido', 'id_usuario', 'fecha_hora_pedido', 'pais', 'dispositivo', 'fuente_referencia', 'nombre_producto', 'categoria_producto', 'cantidad', 'precio_unitario', 'monto_descuento', 'monto_total', 'es_devolucion', 'monto_calculado', 'diferencia', 'monto_inconsistente', 'outlier_cantidad', 'dato_incompleto']
id_pedido               0
id_usuario              0
fecha_hora_pedido       0
pais                    0
dispositivo             0
fuente_referencia       0
nombre_producto        30
categoria_producto     30
cantidad               50
precio_unitario        50
monto_descuento        50
monto_total             0
es_devolucion           0
monto_calculado        50
diferencia             50
monto_inconsistente     0
outlier_cantidad        0
dato_incompleto         0
dtype: int64
(7, 4)
(1620, 5)
fecha         0
pais          0
id_campaña    0
canal         0
gasto         0
dtype: int64


## 📋 Resumen de limpieza — rappiplus_orders_raw

**Registros originales:** 25,100 filas, 12 columnas
**Registros finales:** 25,000 filas, 18 columnas
(100 duplicados exactos eliminados; 6 columnas nuevas creadas como banderas de auditoría)

### 1. Fechas
`fecha_hora_pedido` convertida de texto a `datetime`. 0 nulos tras la conversión.
Rango válido: 2025-01-01 a 2025-06-30.

### 2. Variables numéricas
- Se identificaron 4 pedidos con `cantidad` negativa, correspondientes a **devoluciones**
  (evento de negocio real, no error). Marcados en la columna `es_devolucion`, no eliminados.
- Se detectaron 10 pedidos con `cantidad` de 10,000 o 20,000 unidades — valores no
  plausibles para un pedido de delivery. Patrón consistente con error de captura
  (posible dígito de más). Marcados en `outlier_cantidad`, excluidos de KPIs de
  revenue hasta validar con el negocio.

### 3. Consistencia de montos
Diferencias iniciales de ~$0.01 en gran parte del dataset se debían a redondeo del
sistema de origen, no a errores. Al investigar los casos con diferencia mayor
($0.02+), se encontraron 3 filas correspondientes a devoluciones, donde el
`monto_descuento` se aplica de forma distinta (se suma en vez de restar cuando
`cantidad` es negativa). Fórmula corregida:

`monto_calculado = precio_unitario × cantidad − (monto_descuento × signo(cantidad))`

Tras la corrección: **0 filas inconsistentes** (columna `monto_inconsistente`).

### 4. Duplicados
100 filas exactamente duplicadas (mismo `id_pedido` y mismos datos). Eliminadas
con `drop_duplicates()`. No se encontraron casos de `id_pedido` repetido con
datos distintos.

### 5. Variables categóricas
- `pais`: estandarizado (Argentina/argentina → Argentina, mismo criterio para
  México y Colombia).
- `categoria_producto` en `catalog`: se removieron tildes para que coincida con
  el formato de `orders` ("Electrónica" → "Electronica"), habilitando merges
  correctos entre ambos datasets.
- `categoria_producto` en `orders`: se recuperaron 50 categorías faltantes
  cruzando `nombre_producto` contra `catalog` (nulos bajaron de 80 → 30).
  Las 30 filas restantes tampoco tienen `nombre_producto`, por lo que son
  irrecuperables y quedan documentadas como tales.

### 6. Valores nulos
- `dispositivo`, `fuente_referencia`, `pais`: nulos de bajo impacto financiero,
  rellenados como `"Desconocido"`.
- `nombre_producto` (30), `categoria_producto` (30), `cantidad` (50),
  `precio_unitario` (50), `monto_descuento` (50): **no se imputaron**
  (inventar estos valores distorsionaría el cálculo de revenue). 80 filas
  quedaron marcadas en `dato_incompleto` y se excluyen de los KPIs
  financieros del Paso 2.

### Columnas de auditoría creadas
| Columna | Significado |
|---|---|
| `es_devolucion` | `cantidad` negativa → pedido de devolución |
| `outlier_cantidad` | `cantidad` > 10, sospechoso de error de captura |
| `dato_incompleto` | falta `nombre_producto`, `cantidad` o `precio_unitario` |
| `monto_inconsistente` | diferencia entre `monto_total` y `monto_calculado` mayor a $0.02 |
| `monto_calculado` / `diferencia` | valores de soporte para el punto 3 |

---
**📦 Exportación**: Una vez finalizada la limpieza, se exportan los datasets para utilizarlos en la última etapa del proyecto.

In [31]:
# exportar datasets
orders.to_csv('orders_clean.csv', index=False)
catalog.to_csv('catalog_clean.csv', index=False)
marketing.to_csv('marketing_clean.csv', index=False)

In [32]:
for archivo, esperado in [('orders_clean.csv', orders.shape), ('catalog_clean.csv', catalog.shape), ('marketing_clean.csv', marketing.shape)]:
    prueba = pd.read_csv(archivo)
    estado = '✅' if prueba.shape == esperado else '❌'
    print(f'{archivo}: {prueba.shape} (esperado {esperado}) {estado}')

orders_clean.csv: (25000, 18) (esperado (25000, 18)) ✅
catalog_clean.csv: (7, 4) (esperado (7, 4)) ✅
marketing_clean.csv: (1620, 5) (esperado (1620, 5)) ✅


---

## 🔹 Paso 2: Analizar si el negocio es rentable

### 2.1 Cálculo de KPIs principales

**🎯 Objetivo:** Calcular los indicadores clave del negocio para evaluar ingresos, costos y rentabilidad.

Se usan los 3 datasets (`orders`, `catalog`, `marketing_spend`):

**📊 Parte 1: Rentabilidad del negocio**
- ¿Cuál es el ingreso total (revenue)? 
- ¿Cuál es el costo total? 
- ¿Cuánto se ha invertido en marketing? 
- ¿El negocio es rentable? (calcular profit)

---

**📈 Parte 2: Comportamiento de ventas**
- ¿Cuál es el ticket promedio por orden? 
- ¿Cuál es la cantidad promedio de productos por orden? 
- ¿Cuál es el producto más vendido?
- ¿Cuánto se ha gastado en marketing por canal? 

*¿Cuál es el ingreso total (revenue)?*

In [33]:
revenue_total = orders['monto_total'].sum()
print(f'Revenue total: ${revenue_total:,.2f}')

Revenue total: $51,987,462.26


*¿Cuál es el costo total?*

In [34]:
# Unir orders con catalog, usando nombre_producto como llave de conexión
orders_con_costo = orders.merge(
    catalog[['nombre_producto', 'costo_unitario']],
    on='nombre_producto',
    how='left'
)

# Calcular el costo de cada pedido: costo unitario x cantidad comprada
orders_con_costo['costo_pedido'] = orders_con_costo['costo_unitario'] * orders_con_costo['cantidad']

costo_total = orders_con_costo['costo_pedido'].sum()
print(f'Costo total: ${costo_total:,.2f}')

Costo total: $43,124,018.41


In [35]:
# verificacion 
print(orders_con_costo['costo_unitario'].isnull().sum(), 'pedidos sin costo calculado (por nombre_producto faltante)')

30 pedidos sin costo calculado (por nombre_producto faltante)


In [36]:
# margen bruto 
margen_bruto_monto = revenue_total - costo_total
print(f"Margen Bruto ($): ${margen_bruto_monto:,.2f}")


Margen Bruto ($): $8,863,443.85


*¿Cuánto se ha invertido en marketing?*

In [37]:
marketing_total = marketing['gasto'].sum()
print(f'Inversión total en marketing: ${marketing_total:,.2f}')

Inversión total en marketing: $2,871,843.53


In [38]:
# Verificacion que no quedo ningun nulo 
print(marketing['gasto'].isnull().sum(), 'valores nulos en gasto')

0 valores nulos en gasto


In [39]:
# calculo de ¿El negocio es rentable? (Profit)
profit_total = revenue_total - costo_total - marketing_total
print(f'Profit total: ${profit_total:,.2f}')

# Margen de profit como porcentaje del revenue (más fácil de interpretar que el número absoluto)
margen_profit = (profit_total / revenue_total) * 100
print(f'Margen de profit: {margen_profit:.1f}%')

Profit total: $5,991,600.32
Margen de profit: 11.5%


In [40]:
# ¿Cuánto pesan los outliers en el revenue y costo total?
revenue_outliers = orders[orders['outlier_cantidad']]['monto_total'].sum()
costo_outliers = orders_con_costo[orders_con_costo['outlier_cantidad']]['costo_pedido'].sum()

print(f'Revenue de pedidos outlier: ${revenue_outliers:,.2f} ({revenue_outliers/revenue_total*100:.1f}% del total)')
print(f'Costo de pedidos outlier: ${costo_outliers:,.2f} ({costo_outliers/costo_total*100:.1f}% del total)')

Revenue de pedidos outlier: $42,344,700.00 (81.5% del total)
Costo de pedidos outlier: $39,295,200.00 (91.1% del total)


*Piensa en la magnitud de esto: tienes 24,990 pedidos "normales" generando apenas 18.5% del revenue ($51.9M × 0.185 ≈ $9.6M), mientras que 10 pedidos con cantidades de 10,000–20,000 unidades generan el resto. Eso no es un negocio de delivery/suscripción real — ningún usuario de RappiPlus compra 20,000 tablets en un solo pedido. Esto confirma con mucha fuerza la hipótesis que ya habíamos planteado en el Paso 1: son errores de captura (probablemente un par de ceros de más), no ventas reales.
Por qué esto cambia todo tu análisis
Tu profit_total = $5,991,600.32 y margen = 11.5% están calculados sobre datos dominados por un error. Es decir: ese "11.5% de margen" no describe cómo se comporta el negocio real de RappiPlus — describe principalmente 10 filas rotas. Si reportaras este número al jefe tal cual, estarías respondiendo la pregunta de negocio ("¿el negocio es rentable?") con una cifra que en su mayoría no refleja el comportamiento real de los usuarios.
Esto es exactamente el momento en el que, como analista, no avanzas con el número que salió — te detienes, investigas la causa raíz (ya lo hiciste en el Paso 1), y ahora tomas una decisión explícita sobre cómo tratarlo antes de calcular el KPI final.*

In [41]:
# Recalcular revenue, costo y profit EXCLUYENDO los pedidos marcados como outlier
orders_sin_outliers = orders[~orders['outlier_cantidad']]
orders_con_costo_sin_outliers = orders_con_costo[~orders_con_costo['outlier_cantidad']]

revenue_real = orders_sin_outliers['monto_total'].sum()
costo_real = orders_con_costo_sin_outliers['costo_pedido'].sum()
profit_real = revenue_real - costo_real - marketing_total
margen_real = (profit_real / revenue_real) * 100

print(f'Revenue real (sin outliers): ${revenue_real:,.2f}')
print(f'Costo real (sin outliers): ${costo_real:,.2f}')
print(f'Profit real (sin outliers): ${profit_real:,.2f}')
print(f'Margen real: {margen_real:.1f}%')




Revenue real (sin outliers): $9,642,762.26
Costo real (sin outliers): $3,828,818.41
Profit real (sin outliers): $2,942,100.32
Margen real: 30.5%


In [42]:
print(f'\n--- Comparación ---')
print(f'Profit CON outliers:  ${profit_total:,.2f} (margen {margen_profit:.1f}%)')
print(f'Profit SIN outliers:  ${profit_real:,.2f} (margen {margen_real:.1f}%)')


--- Comparación ---
Profit CON outliers:  $5,991,600.32 (margen 11.5%)
Profit SIN outliers:  $2,942,100.32 (margen 30.5%)


In [43]:
# que tan rentables son los outliers
profit_outliers = revenue_outliers - costo_outliers
margen_outliers = (profit_outliers / revenue_outliers) * 100
print(f'Profit de los outliers (antes de marketing): ${profit_outliers:,.2f}')
print(f'Margen de los outliers: {margen_outliers:.1f}%')


Profit de los outliers (antes de marketing): $3,049,500.00
Margen de los outliers: 7.2%


### resumen de datos encontrados ###
Se identificó que 10 pedidos (0.04% del total) representan el 81.5% del revenue y 91.1% del costo total, con cantidades de 10,000–20,000 unidades — no plausibles para un pedido individual de delivery. Excluyendo estos registros (ya marcados como outlier_cantidad en el Paso 1), el margen de profit real del negocio es del 30.5%, frente al 11.5% que se obtiene si se incluyen. Se recomienda validar estos 10 pedidos con el equipo de operaciones antes de decidir si se corrigen o se descartan definitivamente.


*¿Cuál es el ticket promedio por orden?*

In [44]:
ticket_promedio = orders_sin_outliers['monto_total'].mean()
print(f'Ticket promedio: ${ticket_promedio:,.2f}')
ticket_mediana = orders_sin_outliers['monto_total'].median()
print(f'Ticket mediana: ${ticket_mediana:,.2f}')

Ticket promedio: $385.86
Ticket mediana: $341.61


*¿Cuál es la cantidad promedio de productos por orden?*

In [45]:
cantidad_promedio = orders_sin_outliers['cantidad'].mean()
print(f'Cantidad promedio de productos por orden: {cantidad_promedio:.2f}')

Cantidad promedio de productos por orden: 1.50


*¿Cuál es la cantidad promedio de productos por orden?*

In [46]:
# Versión alternativa: solo compras reales, sin devoluciones
cantidad_promedio_compras = orders_sin_outliers[~orders_sin_outliers['es_devolucion']]['cantidad'].mean()
print(f'Cantidad promedio (solo compras, sin devoluciones): {cantidad_promedio_compras:.2f}')

Cantidad promedio (solo compras, sin devoluciones): 1.50


*¿Cuál es el producto más vendido?*

In [47]:
# Por CANTIDAD total de unidades vendidas
mas_vendido_cantidad = orders_sin_outliers.groupby('nombre_producto')['cantidad'].sum().sort_values(ascending=False)
print('Producto más vendido (por unidades):')
print(mas_vendido_cantidad.head())

Producto más vendido (por unidades):
nombre_producto
Vacuum-Pro-Black      6284.0
Blender-XL-Red        6279.0
Jacket-Winter-M       6256.0
Sneakers-Urban-42     6172.0
Laptop-Gaming-16GB    4198.0
Name: cantidad, dtype: float64


In [48]:
# Por NÚMERO DE PEDIDOS en los que aparece (popularidad, no volumen)
mas_vendido_pedidos = orders_sin_outliers['nombre_producto'].value_counts()
print('\nProducto más vendido (por número de pedidos):')
print(mas_vendido_pedidos.head())


Producto más vendido (por número de pedidos):
Blender-XL-Red        4179
Jacket-Winter-M       4176
Vacuum-Pro-Black      4176
Sneakers-Urban-42     4148
Laptop-Gaming-16GB    2772
Name: nombre_producto, dtype: int64


*¿Cuánto se ha gastado en marketing por canal?*

In [49]:
gasto_por_canal = marketing.groupby('canal')['gasto'].sum().sort_values(ascending=False)
print(gasto_por_canal)


canal
social         976818.37
organic        972650.96
paid_search    922374.20
Name: gasto, dtype: float64


---

In [50]:
gasto_por_pais_canal = marketing.groupby(['pais', 'canal'])['gasto'].sum().sort_values(ascending=False)
print('\nGasto por país y canal:')
print(gasto_por_pais_canal)


Gasto por país y canal:
pais       canal      
Mexico     social         332113.79
Argentina  social         331175.63
Mexico     organic        330231.46
           paid_search    326150.26
Colombia   organic        323222.51
Argentina  organic        319196.99
Colombia   social         313528.95
           paid_search    298901.96
Argentina  paid_search    297321.98
Name: gasto, dtype: float64


In [51]:
# En Python, para tener el número de referencia correcto
orders['mes'] = orders['fecha_hora_pedido'].dt.to_period('M')
print(orders[~orders['outlier_cantidad']].groupby('mes')['monto_total'].sum())

mes
2025-01    1659370.12
2025-02    1444109.37
2025-03    1662993.09
2025-04    1602801.27
2025-05    1654701.34
2025-06    1618787.07
Freq: M, Name: monto_total, dtype: float64






---

### Parte 1: Rentabilidad del negocio

| KPI | Valor (con outliers) | Valor (sin outliers) |
|---|---|---|
| Revenue total | $51,987,462.26 | $9,642,762.26 |
| Costo total | $43,124,018.41 | $3,828,818.41 |
| Inversión en marketing | $2,872,881.54 | $2,872,881.54 |
| **Profit** | $5,991,600.32 | **$2,942,100.32** |
| **Margen de profit** | 11.5% | **30.5%** |

**🔎 Hallazgo crítico:** 10 pedidos (0.04% del dataset) con `cantidad` de 10,000–20,000 
unidades —no plausibles para un pedido individual de delivery— concentran el 81.5% 
del revenue y el 91.1% del costo total. Al incluirlos, distorsionan el margen real 
del negocio hacia abajo (11.5% vs. el 30.5% real), porque su propio margen individual 
(~7.2%) es menor al del negocio típico. 

**Conclusión de negocio:** el margen real de RappiPlus es del **30.5%**, no del 11.5%. 
Se recomienda validar estos 10 pedidos con el equipo de operaciones antes de decidir 
si se corrigen o se descartan definitivamente del histórico.

**Cómo se calculó el costo:** se unió `orders` con `catalog` mediante un `merge` 
(`on='nombre_producto', how='left'`) para traer `costo_unitario`, y se multiplicó 
por `cantidad`. 30 pedidos quedaron sin costo calculable (los mismos que no tienen 
`nombre_producto`, ya documentados en el Paso 1 como irrecuperables).

---

### Parte 2: Comportamiento de ventas (excluyendo outliers)

| KPI | Valor | Nota |
|---|---|---|
| Ticket promedio | $385.86 | Mediana: $341.61 — la diferencia indica asimetría hacia la derecha (algunos pedidos altos "jalan" el promedio) |
| Cantidad promedio por orden | 1.50 productos | Sin diferencia relevante al excluir devoluciones (solo 4 filas) |
| Producto más vendido (unidades) | Vacuum-Pro-Black (6,284) | Casi empatado con Blender-XL-Red (6,279) |
| Producto más vendido (frecuencia) | Blender-XL-Red (4,179 pedidos) | Casi empatado con Vacuum-Pro-Black (4,176) |
| Producto menos vendido | Laptop-Gaming-16GB | Consistente en ambas métricas |
| Gasto en marketing por canal | social $976,818 / organic $972,651 / paid_search $922,374 | Ver hallazgo de corrección abajo |

**Nota:** "producto más vendido" depende de cómo se mida — por volumen de unidades 
(Vacuum-Pro-Black) o por número de pedidos distintos en los que aparece 
(Blender-XL-Red). Ambos están prácticamente empatados; se reportan los dos en 
lugar de forzar un único ganador.

---

### 🔧 Corrección aplicada durante este paso (retroactiva al Paso 1)

Se detectó que la columna `canal` en `marketing` tenía un error de origen: al 
recuperar valores nulos desde `id_campaña` usando `.str.split('_').str[0]`, el 
canal `paid_search` se partía incorrectamente en `"paid"`, generando un cuarto 
canal inexistente con $59,285.99 mal clasificados.

**Corrección:** se reemplazó `.split()` por `.rsplit('_', n=1)`, que separa desde 
la derecha y respeta el guion bajo interno de `paid_search`. Se corrigió en el 
origen (Paso 1) y se re-exportó `marketing_clean.csv`, en vez de dejar el ajuste 
solo como parche en este paso — así el dataset queda correcto para cualquier 
análisis futuro que lo use.

**Verificación:** `922,374.20 = 863,088.21 (paid_search original) + 59,285.99 
(mal clasificado como "paid")` ✅

## 🔹 Paso 3: Entender dónde se pierden los usuarios (funnel de conversión)

**🎯 Objetivo:** Analizar el comportamiento de los usuarios para identificar en qué etapa del proceso se pierden.


⚙️**Conexión a la base de datos**:  
Se ejecuta la línea de configuración para conectar con la base de datos y aplicar consultas SQL en la tabla **events**.

---

**📊 Parte 1: Construcción del funnel**
- ¿Cuántos usuarios llegan a cada etapa del funnel?  
- Se calcula el número de usuarios únicos por `nombre_evento`  
- Se ordenan los eventos según el flujo del usuario  

---

**📉 Parte 2: Análisis de conversión**
- Se calcula la tasa de conversión entre cada paso del funnel  
- Se identifica en qué etapa se pierde la mayor cantidad de usuarios  
- ¿Cuál es la tasa de conversión final?
---

In [52]:
import pandas as pd
from sqlalchemy import create_engine

# ======================
# Conexión (NO modificar)
# ======================
db_config = {
    'user': 'practicum_student',
    'pwd': 'QnmDH8Sc2TQLvy2G3Vvh7',
    'host': 'yp-trainers-practicum.cluster-czs0gxyx2d8w.us-east-1.rds.amazonaws.com',
    'port': 5432,
    'db': 'data-analyst-production-db-en'
}

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
    db_config['user'],
    db_config['pwd'],
    db_config['host'],
    db_config['port'],
    db_config['db']
)

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [53]:
# Explorar tabla events
# =========================
query_events = '''
SELECT *
FROM events;
'''
events = pd.read_sql(query_events, con=engine)
events.head(7)

,id_usuario,id_sesion,nombre_evento,timestamp_evento,pais,dispositivo,fuente_referencia,categoria_producto
0,user_6772,6a97f2af-32ae-4186-8c92-04025be1a27b,first_visit,2025-05-17,Colombia,desktop,organic,Moda
1,user_5883,369b767c-1c33-4b2f-a652-c7c0ef92cfc9,add_to_cart,2025-02-23,Mexico,mobile,social,Hogar
2,user_5946,60039041-e78b-474c-87b3-c0b7e9c30708,add_payment_info,2025-05-15,Colombia,desktop,social,Electronica
3,user_827,18252a64-f389-4ef7-9e58-dadad4a3491e,purchase,2025-03-31,Mexico,mobile,social,Moda
4,user_2361,221b364e-cdc5-4668-b698-18d5ba849a67,first_visit,2025-01-22,Argentina,desktop,paid_search,Electronica
5,user_4402,0892643c-f80d-4eb4-9ca8-3638b16c6ed4,begin_checkout,2025-04-21,Argentina,desktop,paid_search,Hogar
6,user_767,0005a13c-3c1f-496b-a26a-f3c282f19e13,begin_checkout,2025-03-01,Mexico,mobile,social,Electronica


In [54]:
print(events.shape)
print(events.info())

(120000, 8)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 120000 entries, 0 to 119999
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   id_usuario          120000 non-null  object
 1   id_sesion           120000 non-null  object
 2   nombre_evento       120000 non-null  object
 3   timestamp_evento    120000 non-null  object
 4   pais                120000 non-null  object
 5   dispositivo         120000 non-null  object
 6   fuente_referencia   120000 non-null  object
 7   categoria_producto  120000 non-null  object
dtypes: object(8)
memory usage: 7.3+ MB
None


In [55]:
query_total = '''
SELECT COUNT(*) AS total_registros
FROM events;
'''
total = pd.read_sql(query_total, con=engine)
print(total)

   total_registros
0           120000


In [56]:
query_categorias = '''
SELECT DISTINCT nombre_evento
FROM events;
'''
categorias_evento = pd.read_sql(query_categorias, con=engine)
print(categorias_evento)

      nombre_evento
0  add_payment_info
1       first_visit
2    begin_checkout
3       add_to_cart
4       select_item
5          purchase


In [57]:
query_fuente = 'SELECT DISTINCT fuente_referencia FROM events;'
fuente_unica = pd.read_sql(query_fuente, con=engine)
print(fuente_unica)

query_categoria_prod = 'SELECT DISTINCT categoria_producto FROM events;'
categoria_prod_unica = pd.read_sql(query_categoria_prod, con=engine)
print(categoria_prod_unica)

  fuente_referencia
0       paid_search
1           organic
2            social
  categoria_producto
0               Moda
1        Electronica
2              Hogar


In [58]:
# PARTE 1: Totales del funnel
# ======================

query_totales = '''
SELECT 
    nombre_evento,
    COUNT(DISTINCT id_usuario) AS usuarios_unicos
FROM events
GROUP BY nombre_evento
ORDER BY usuarios_unicos DESC;
'''
totales = pd.read_sql(query_totales, con=engine)
print(totales)

      nombre_evento  usuarios_unicos
0       first_visit             7796
1       add_to_cart             7634
2       select_item             7582
3    begin_checkout             7208
4  add_payment_info             6250
5          purchase             6240


In [59]:
query_totales = '''
SELECT 
    nombre_evento,
    COUNT(DISTINCT id_usuario) AS usuarios_unicos,
    CASE nombre_evento
        WHEN 'first_visit' THEN 1
        WHEN 'select_item' THEN 2
        WHEN 'add_to_cart' THEN 3
        WHEN 'begin_checkout' THEN 4
        WHEN 'add_payment_info' THEN 5
        WHEN 'purchase' THEN 6
    END AS orden_funnel
FROM events
GROUP BY nombre_evento
ORDER BY orden_funnel;
'''
totales = pd.read_sql(query_totales, con=engine)
print(totales)

      nombre_evento  usuarios_unicos  orden_funnel
0       first_visit             7796             1
1       select_item             7582             2
2       add_to_cart             7634             3
3    begin_checkout             7208             4
4  add_payment_info             6250             5
5          purchase             6240             6


In [60]:
# ¿Cuántos usuarios tienen add_to_cart PERO nunca tuvieron select_item?
query_investigacion = '''
SELECT COUNT(DISTINCT id_usuario) AS usuarios_sin_select_item
FROM events
WHERE nombre_evento = 'add_to_cart'
AND id_usuario NOT IN (
    SELECT id_usuario FROM events WHERE nombre_evento = 'select_item'
);
'''
investigacion = pd.read_sql(query_investigacion, con=engine)
print(investigacion)

   usuarios_sin_select_item
0                       401


In [61]:
# PARTE 2: Conversiones
# ======================


query_conversion = '''
WITH totales AS (
    SELECT 
        nombre_evento,
        COUNT(DISTINCT id_usuario) AS usuarios_unicos,
        CASE nombre_evento
            WHEN 'first_visit' THEN 1
            WHEN 'select_item' THEN 2
            WHEN 'add_to_cart' THEN 3
            WHEN 'begin_checkout' THEN 4
            WHEN 'add_payment_info' THEN 5
            WHEN 'purchase' THEN 6
        END AS orden_funnel
    FROM events
    GROUP BY nombre_evento
)
SELECT
    nombre_evento,
    orden_funnel,
    usuarios_unicos,
    LAG(usuarios_unicos) OVER (ORDER BY orden_funnel) AS usuarios_etapa_anterior,
    ROUND(
        100.0 * usuarios_unicos / LAG(usuarios_unicos) OVER (ORDER BY orden_funnel),
        1
    ) AS conversion_pct
FROM totales
ORDER BY orden_funnel;
'''

conversion = pd.read_sql(query_conversion, con=engine)
print(conversion)

      nombre_evento  orden_funnel  usuarios_unicos  usuarios_etapa_anterior  \
0       first_visit             1             7796                      NaN   
1       select_item             2             7582                   7796.0   
2       add_to_cart             3             7634                   7582.0   
3    begin_checkout             4             7208                   7634.0   
4  add_payment_info             5             6250                   7208.0   
5          purchase             6             6240                   6250.0   

   conversion_pct  
0             NaN  
1            97.3  
2           100.7  
3            94.4  
4            86.7  
5            99.8  


In [62]:
conversion_final = (6240 / 7796) * 100
print(f'{conversion_final:.1f}%')

80.0%


In [63]:
# ¿fuente_referencia tiene el mismo valor para TODOS los eventos de un mismo usuario,
# o podría variar entre eventos de la misma persona?
query_check = '''
SELECT id_usuario, COUNT(DISTINCT fuente_referencia) AS fuentes_distintas
FROM events
GROUP BY id_usuario
HAVING COUNT(DISTINCT fuente_referencia) > 1
LIMIT 5;
'''
check = pd.read_sql(query_check, con=engine)
print(check)

  id_usuario  fuentes_distintas
0     user_0                  3
1     user_1                  3
2    user_10                  3
3   user_100                  3
4  user_1000                  3


In [64]:
query_extension = '''
SELECT 
    COUNT(DISTINCT id_usuario) AS usuarios_con_multiples_fuentes
FROM (
    SELECT id_usuario
    FROM events
    GROUP BY id_usuario
    HAVING COUNT(DISTINCT fuente_referencia) > 1
) AS sub;
'''
extension = pd.read_sql(query_extension, con=engine)
print(extension)

query_total_usuarios = 'SELECT COUNT(DISTINCT id_usuario) AS total_usuarios FROM events;'
total_usuarios = pd.read_sql(query_total_usuarios, con=engine)
print(total_usuarios)

   usuarios_con_multiples_fuentes
0                            7999
   total_usuarios
0            8000


In [65]:
query_first_touch = '''
SELECT 
    id_usuario,
    fuente_referencia AS canal_primer_contacto
FROM events
WHERE nombre_evento = 'first_visit';
'''
first_touch = pd.read_sql(query_first_touch, con=engine)
print(first_touch.shape)
print(first_touch.head())

(29957, 2)
  id_usuario canal_primer_contacto
0  user_6772               organic
1  user_2361           paid_search
2  user_4897                social
3  user_6121               organic
4  user_5462                social


In [66]:
print(first_touch['id_usuario'].duplicated().sum(), 'usuarios con más de un first_visit')

22161 usuarios con más de un first_visit


In [67]:
query_check_fecha = '''
SELECT id_usuario, timestamp_evento
FROM events
WHERE nombre_evento = 'first_visit'
ORDER BY id_usuario
LIMIT 5;
'''
check_fecha = pd.read_sql(query_check_fecha, con=engine)
print(check_fecha)
print(check_fecha.dtypes)

  id_usuario timestamp_evento
0     user_0       2025-04-19
1     user_0       2025-01-24
2     user_1       2025-04-20
3    user_10       2025-03-16
4    user_10       2025-04-29
id_usuario          object
timestamp_evento    object
dtype: object


In [68]:
query_first_touch = '''
WITH visitas_numeradas AS (
    SELECT 
        id_usuario,
        fuente_referencia,
        timestamp_evento,
        ROW_NUMBER() OVER (
            PARTITION BY id_usuario 
            ORDER BY timestamp_evento ASC
        ) AS numero_visita
    FROM events
    WHERE nombre_evento = 'first_visit'
)
SELECT 
    id_usuario,
    fuente_referencia AS canal_primer_contacto,
    timestamp_evento AS fecha_primera_visita
FROM visitas_numeradas
WHERE numero_visita = 1;
'''
first_touch = pd.read_sql(query_first_touch, con=engine)
print(first_touch.shape)
print(first_touch.head())

(7796, 3)
  id_usuario canal_primer_contacto fecha_primera_visita
0     user_0                social           2025-01-24
1     user_1               organic           2025-04-20
2    user_10               organic           2025-02-20
3   user_100                social           2025-03-21
4  user_1000                social           2025-02-10


In [69]:
print(first_touch.shape)
print(first_touch['id_usuario'].duplicated().sum(), 'usuarios con más de una fila (debería ser 0)')

(7796, 3)
0 usuarios con más de una fila (debería ser 0)


## funnel segmentado por canal ##

In [70]:
#traer todos los eventos (sin fuente_referencia, porque ya no la vamos a usar — usamos el canal fijo)
query_eventos_completos = '''
SELECT id_usuario, nombre_evento
FROM events;
'''
eventos = pd.read_sql(query_eventos_completos, con=engine)
print(eventos.shape)

(120000, 2)


In [71]:
# paso 2 unir cada evento con el canal de adquisición ORIGINAL de ese usuario
eventos_con_canal = eventos.merge(first_touch[['id_usuario', 'canal_primer_contacto']], on='id_usuario', how='left')
print(eventos_con_canal.head())
print(eventos_con_canal['canal_primer_contacto'].isnull().sum(), 'eventos sin canal de adquisición asignado')

  id_usuario     nombre_evento canal_primer_contacto
0  user_6772       first_visit                social
1  user_5883       add_to_cart           paid_search
2  user_5946  add_payment_info               organic
3   user_827          purchase           paid_search
4  user_2361       first_visit           paid_search
2282 eventos sin canal de adquisición asignado


In [72]:
# paso 3 el funnel segmentado, ya con el canal correcto
funnel_por_canal = eventos_con_canal.groupby(['canal_primer_contacto', 'nombre_evento'])['id_usuario'].nunique().reset_index()
funnel_por_canal.columns = ['canal_primer_contacto', 'nombre_evento', 'usuarios_unicos']

orden_map = {'first_visit':1, 'select_item':2, 'add_to_cart':3, 'begin_checkout':4, 'add_payment_info':5, 'purchase':6}
funnel_por_canal['orden_funnel'] = funnel_por_canal['nombre_evento'].map(orden_map)

funnel_por_canal = funnel_por_canal.sort_values(['canal_primer_contacto', 'orden_funnel']).reset_index(drop=True)
funnel_por_canal['usuarios_etapa_anterior'] = funnel_por_canal.groupby('canal_primer_contacto')['usuarios_unicos'].shift(1)
funnel_por_canal['conversion_pct'] = (funnel_por_canal['usuarios_unicos'] / funnel_por_canal['usuarios_etapa_anterior']) * 100

print(funnel_por_canal)

   canal_primer_contacto     nombre_evento  usuarios_unicos  orden_funnel  \
0                organic       first_visit             2547             1   
1                organic       select_item             2417             2   
2                organic       add_to_cart             2406             3   
3                organic    begin_checkout             2304             4   
4                organic  add_payment_info             1976             5   
5                organic          purchase             1998             6   
6            paid_search       first_visit             2591             1   
7            paid_search       select_item             2448             2   
8            paid_search       add_to_cart             2487             3   
9            paid_search    begin_checkout             2336             4   
10           paid_search  add_payment_info             2026             5   
11           paid_search          purchase             1997             6   

In [73]:
conversion_final_canal = funnel_por_canal[funnel_por_canal['nombre_evento'].isin(['first_visit', 'purchase'])].pivot(
    index='canal_primer_contacto', columns='nombre_evento', values='usuarios_unicos'
)
conversion_final_canal['conversion_total_pct'] = (conversion_final_canal['purchase'] / conversion_final_canal['first_visit']) * 100
print(conversion_final_canal)

nombre_evento          first_visit  purchase  conversion_total_pct
canal_primer_contacto                                             
organic                       2547      1998             78.445230
paid_search                   2591      1997             77.074489
social                        2658      2073             77.990971


In [74]:
query_check_categoria = '''
SELECT 
    nombre_evento, 
    COUNT(*) AS total, 
    COUNT(categoria_producto) AS con_categoria
FROM events
GROUP BY nombre_evento;
'''
check_categoria = pd.read_sql(query_check_categoria, con=engine)
print(check_categoria)

      nombre_evento  total  con_categoria
0  add_payment_info  12018          12018
1       first_visit  29957          29957
2    begin_checkout  17971          17971
3       add_to_cart  24157          24157
4       select_item  23887          23887
5          purchase  12010          12010


In [75]:
query_funnel_categoria = '''
SELECT 
    categoria_producto,
    nombre_evento,
    COUNT(DISTINCT id_usuario) AS usuarios_unicos,
    CASE nombre_evento
        WHEN 'first_visit' THEN 1
        WHEN 'select_item' THEN 2
        WHEN 'add_to_cart' THEN 3
        WHEN 'begin_checkout' THEN 4
        WHEN 'add_payment_info' THEN 5
        WHEN 'purchase' THEN 6
    END AS orden_funnel
FROM events
GROUP BY categoria_producto, nombre_evento
ORDER BY categoria_producto, orden_funnel;
'''
funnel_categoria = pd.read_sql(query_funnel_categoria, con=engine)
print(funnel_categoria)

   categoria_producto     nombre_evento  usuarios_unicos  orden_funnel
0         Electronica       first_visit             5680             1
1         Electronica       select_item             5007             2
2         Electronica       add_to_cart             5126             3
3         Electronica    begin_checkout             4227             4
4         Electronica  add_payment_info             3178             5
5         Electronica          purchase             3177             6
6               Hogar       first_visit             5695             1
7               Hogar       select_item             5073             2
8               Hogar       add_to_cart             5014             3
9               Hogar    begin_checkout             4234             4
10              Hogar  add_payment_info             3139             5
11              Hogar          purchase             3203             6
12               Moda       first_visit             5694             1
13    

In [76]:
funnel_categoria = funnel_categoria.sort_values(['categoria_producto', 'orden_funnel']).reset_index(drop=True)
funnel_categoria['usuarios_etapa_anterior'] = funnel_categoria.groupby('categoria_producto')['usuarios_unicos'].shift(1)
funnel_categoria['conversion_pct'] = (funnel_categoria['usuarios_unicos'] / funnel_categoria['usuarios_etapa_anterior']) * 100
print(funnel_categoria)

   categoria_producto     nombre_evento  usuarios_unicos  orden_funnel  \
0         Electronica       first_visit             5680             1   
1         Electronica       select_item             5007             2   
2         Electronica       add_to_cart             5126             3   
3         Electronica    begin_checkout             4227             4   
4         Electronica  add_payment_info             3178             5   
5         Electronica          purchase             3177             6   
6               Hogar       first_visit             5695             1   
7               Hogar       select_item             5073             2   
8               Hogar       add_to_cart             5014             3   
9               Hogar    begin_checkout             4234             4   
10              Hogar  add_payment_info             3139             5   
11              Hogar          purchase             3203             6   
12               Moda       first_visi

Aquí tienes el Markdown completo del Paso 3, con la misma estructura que usamos en el Paso 1 y el Paso 2:
markdown## 📋 Resumen — Paso 3: Funnel de conversión (tabla `events`)

**Fuente de datos:** tabla `events` en base de datos PostgreSQL (conexión vía SQLAlchemy)
**Registros totales:** 120,000 eventos, 8,000 usuarios únicos, 6 tipos de evento

---

### Parte 1: Funnel general

Se calculó el número de usuarios únicos (`COUNT(DISTINCT id_usuario)`) por cada 
etapa del proceso de compra, ordenado según la lógica de negocio (no según 
volumen), usando un `CASE` en SQL para asignar la posición correcta:

| Etapa | Usuarios únicos | Conversión desde la etapa anterior |
|---|---|---|
| first_visit | 7,796 | — |
| select_item | 7,582 | 97.3% |
| add_to_cart | 7,634 | **100.7%** ⚠️ |
| begin_checkout | 7,208 | 94.4% |
| add_payment_info | 6,250 | **86.7%** ← mayor caída |
| purchase | 6,240 | 99.8% |

**Conversión final (first_visit → purchase): 80.0%**

**🔎 Hallazgo — conversión superior al 100%:** la etapa `add_to_cart` muestra 
más usuarios que `select_item`, lo cual es matemáticamente imposible en un 
funnel estrictamente secuencial. Se investigó y se confirmó: 401 usuarios 
llegaron a `add_to_cart` sin tener nunca un evento `select_item` registrado, 
lo que indica una **ruta de conversión alterna** (ej. agregar al carrito 
directo desde una recomendación o notificación), no un error de captura. 
Este patrón se repite de forma consistente en los desgloses por canal y por 
categoría (ver abajo), confirmando que es estructural, no aislado.

**🔎 Hallazgo — etapa más crítica:** `add_payment_info` concentra la mayor 
pérdida de usuarios (13.3%). Como esta caída es similar en los 3 canales de 
adquisición, el problema parece estar en el **proceso de pago en sí** 
(fricción del formulario, falta de métodos de pago), no en la calidad de 
los usuarios que trae un canal específico.

---

### Parte 2: Funnel segmentado por canal de adquisición

**Ajuste metodológico importante:** la columna `fuente_referencia` de `events` 
**no es un atributo fijo del usuario** — el 99.99% de los usuarios (7,999 de 
8,000) tienen múltiples fuentes distintas a lo largo de sus eventos, lo cual 
haría inválida cualquier segmentación directa por esa columna.

**Solución aplicada — first-touch attribution:** se identificó el canal de 
adquisición real de cada usuario usando `ROW_NUMBER() OVER (PARTITION BY 
id_usuario ORDER BY timestamp_evento ASC)`, quedándose con el evento 
`first_visit` más antiguo de cada usuario (se descubrió que `first_visit` se 
registra por sesión, no una sola vez por usuario: 22,161 usuarios tenían más 
de uno). Este canal fijo se unió al resto de sus eventos mediante `merge`.

| Canal | first_visit | purchase | Conversión total |
|---|---|---|---|
| organic | 2,547 | 1,998 | 78.4% |
| paid_search | 2,591 | 1,997 | 77.1% |
| social | 2,658 | 2,073 | 78.0% |

**Conclusión:** los 3 canales convierten de forma prácticamente idéntica 
(diferencia máxima de 1.3 puntos porcentuales). No hay evidencia de que algún 
canal traiga usuarios de menor calidad — la decisión de inversión en marketing 
(Paso 2) debería basarse más en el costo de adquisición por canal que en la 
tasa de conversión, ya que esta última no discrimina entre canales.

*Nota: 2,282 eventos (de usuarios sin `first_visit` registrado) quedaron sin 
canal de adquisición asignado y se excluyeron de esta segmentación.*

---

### Parte 3: Funnel segmentado por categoría de producto

Se verificó primero que `categoria_producto` no tiene valores nulos en ningún 
evento (incluyendo `first_visit`), por lo que no fue necesario excluir ninguna 
fila antes de segmentar.

| Categoría | first_visit | Conversión más débil | Etapa |
|---|---|---|---|
| Electronica | 5,680 | 75.2% | add_payment_info |
| Hogar | 5,695 | 74.1% | add_payment_info |
| Moda | 5,694 | 75.1% | add_payment_info |

**Conclusión:** las 3 categorías muestran el mismo patrón que el funnel 
general — la mayor caída ocurre en `add_payment_info`, con porcentajes muy 
similares entre sí. Refuerza la conclusión de que el problema es del proceso 
de pago, transversal a toda la plataforma.

**⚠️ Advertencia metodológica:** a diferencia del canal (first-touch), 
`categoria_producto` describe la categoría de **cada evento**, no un atributo 
fijo por usuario. Un mismo usuario puede aparecer en más de una categoría 
(ej. navega Moda en una sesión y Electrónica en otra). Por esto, **la suma de 
usuarios de las 3 categorías en `first_visit` (17,069) no equivale ni debe 
compararse con el total de usuarios únicos de la plataforma (7,796)** — 
representa interacciones por categoría, no usuarios distintos.

---

## 🔹 Paso 4: Evaluar si los usuarios regresan (retención por cohortes)

**🎯 Objetivo:** Analizar la retención de usuarios para entender si regresan después de registrarse.

**Tablas**

- `users` 
- `user_activity` 

---
1. Se identifica la cohorte de cada usuario según el **mes de registro**.


2. Se calcula la retención semanal: cuántos usuarios **se mantienen activos** en cada semana desde su registro.
   - `retenido_w1`: usuarios activos en la semana 1  
   - `retenido_w2`: usuarios activos en la semana 2  
   - `retenido_w3`: usuarios activos en la semana 3  


3. Se calcula el porcentaje de retención para cada semana, dividiendo los usuarios retenidos entre los clientes iniciales de la cohorte:  
   - `semana_1`: porcentaje de usuarios retenidos en la semana 1  
   - `semana_2`: porcentaje de usuarios retenidos en la semana 2  
   - `semana_3`: porcentaje de usuarios retenidos en la semana 3  

Se revisa que la columna de fecha esté en formato correcto (`DATE`).  
Se realiza la conversión usando: `CAST(fecha_registro AS DATE)`

In [77]:
# Explorar tabla users
# =========================
query_users = '''
SELECT *
FROM users;
'''
users = pd.read_sql(query_users, con=engine)
users.head(3)

,id_usuario,fecha_registro,país,dispositivo,tipo_plan
0,user_0,2025-01-29,Mexico,mobile,free
1,user_1,2025-01-07,Mexico,mobile,free
2,user_2,2025-03-12,Argentina,mobile,free


In [78]:
# Explorar tabla user_activity
# =========================
query_user_activity = '''
SELECT *
FROM user_activity;
'''
user_activity = pd.read_sql(query_user_activity, con=engine)
user_activity.head(3)

,id_usuario,fecha_actividad,dias_despues_registro,activo
0,user_0,2025-02-05,7,0
1,user_0,2025-02-12,14,1
2,user_0,2025-02-19,21,1


In [79]:
# ¿Todos los usuarios tienen exactamente 3 filas (una para cada semana: 7, 14, 21)?
print(user_activity.groupby('id_usuario').size().value_counts())

# ¿Qué valores de dias_despues_registro existen en total?
print(user_activity['dias_despues_registro'].unique())

# ¿Qué valores toma la columna activo?
print(user_activity['activo'].unique())

4    8000
dtype: int64
[ 7 14 21 28]
[0 1]


### Paso 1: identificar la cohorte de cada usuario ###

In [80]:
query_cohortes = '''
SELECT 
    id_usuario,
    CAST(fecha_registro AS DATE) AS fecha_registro,
    DATE_TRUNC('month', CAST(fecha_registro AS DATE)) AS mes_cohorte
FROM users;
'''
cohortes = pd.read_sql(query_cohortes, con=engine)
print(cohortes.head())

  id_usuario fecha_registro               mes_cohorte
0     user_0     2025-01-29 2025-01-01 00:00:00+00:00
1     user_1     2025-01-07 2025-01-01 00:00:00+00:00
2     user_2     2025-03-12 2025-03-01 00:00:00+00:00
3     user_3     2025-03-04 2025-03-01 00:00:00+00:00
4     user_4     2025-02-27 2025-02-01 00:00:00+00:00


### Paso 2: unir cohortes con actividad, y calcular retención por semana ###

In [81]:
# Retención por cohortes
# ======================


query_cohort_retention_final = '''
WITH cohortes AS (
    SELECT 
        id_usuario,
        DATE_TRUNC('month', CAST(fecha_registro AS DATE)) AS mes_cohorte
    FROM users
),
tamanos_cohorte AS (
    SELECT 
        mes_cohorte,
        COUNT(DISTINCT id_usuario) AS usuarios_iniciales
    FROM cohortes
    GROUP BY mes_cohorte
),
retencion AS (
    SELECT
        c.mes_cohorte,
        COUNT(DISTINCT CASE WHEN ua.dias_despues_registro = 7 AND ua.activo = 1 THEN ua.id_usuario END) AS retenido_w1,
        COUNT(DISTINCT CASE WHEN ua.dias_despues_registro = 14 AND ua.activo = 1 THEN ua.id_usuario END) AS retenido_w2,
        COUNT(DISTINCT CASE WHEN ua.dias_despues_registro = 21 AND ua.activo = 1 THEN ua.id_usuario END) AS retenido_w3
    FROM cohortes c
    JOIN user_activity ua ON c.id_usuario = ua.id_usuario
    GROUP BY c.mes_cohorte
)
SELECT 
    r.mes_cohorte,
    t.usuarios_iniciales,
    r.retenido_w1,
    r.retenido_w2,
    r.retenido_w3,
    ROUND(100.0 * r.retenido_w1 / t.usuarios_iniciales, 1) AS semana_1,
    ROUND(100.0 * r.retenido_w2 / t.usuarios_iniciales, 1) AS semana_2,
    ROUND(100.0 * r.retenido_w3 / t.usuarios_iniciales, 1) AS semana_3
FROM retencion r
JOIN tamanos_cohorte t ON r.mes_cohorte = t.mes_cohorte
ORDER BY r.mes_cohorte;
'''

# Ejecutar la consulta

cohorte_final = pd.read_sql(query_cohort_retention_final, con=engine)
print(cohorte_final)

                mes_cohorte  usuarios_iniciales  retenido_w1  retenido_w2  \
0 2025-01-01 00:00:00+00:00                1627          697          668   
1 2025-02-01 00:00:00+00:00                1444          611          609   
2 2025-03-01 00:00:00+00:00                1636          677          705   
3 2025-04-01 00:00:00+00:00                1606          680          697   
4 2025-05-01 00:00:00+00:00                1687          695          676   

   retenido_w3  semana_1  semana_2  semana_3  
0          656      42.8      41.1      40.3  
1          635      42.3      42.2      44.0  
2          690      41.4      43.1      42.2  
3          663      42.3      43.4      41.3  
4          706      41.2      40.1      41.8  


In [82]:
query_check_junio = '''
SELECT COUNT(DISTINCT id_usuario) AS usuarios_junio
FROM users
WHERE DATE_TRUNC('month', CAST(fecha_registro AS DATE)) = '2025-06-01';
'''
check_junio = pd.read_sql(query_check_junio, con=engine)
print(check_junio)

   usuarios_junio
0               0


In [83]:
query_rango_registro = '''
SELECT 
    MIN(CAST(fecha_registro AS DATE)) AS primer_registro,
    MAX(CAST(fecha_registro AS DATE)) AS ultimo_registro
FROM users;
'''
rango_registro = pd.read_sql(query_rango_registro, con=engine)
print(rango_registro)

  primer_registro ultimo_registro
0      2025-01-01      2025-05-31


## 📋 Resumen — Paso 4: Retención de usuarios por cohortes

**Tablas utilizadas:** `users`, `user_activity` (base de datos PostgreSQL)
**Metodología:** cohortes mensuales según `fecha_registro`, con retención medida 
en los puntos de chequeo de 7, 14 y 21 días después del registro (`user_activity`).

---

### Identificación de cohortes

Se agrupó a cada usuario según el mes de su `fecha_registro`, usando 
`DATE_TRUNC('month', ...)`. Se aplicó `CAST(fecha_registro AS DATE)` para 
asegurar que la comparación de fechas fuera exacta, sin componente de hora.

### Tabla de retención por cohorte

| Cohorte | Usuarios iniciales | Semana 1 | Semana 2 | Semana 3 |
|---|---|---|---|---|
| Enero 2025 | 1,627 | 42.8% | 41.1% | 40.3% |
| Febrero 2025 | 1,444 | 42.3% | 42.2% | 44.0% |
| Marzo 2025 | 1,636 | 41.4% | 43.1% | 42.2% |
| Abril 2025 | 1,606 | 42.3% | 43.4% | 41.3% |
| Mayo 2025 | 1,687 | 41.2% | 40.1% | 41.8% |

**Validación:** la suma de usuarios iniciales de las 5 cohortes (1,627 + 1,444 + 
1,636 + 1,606 + 1,687 = 8,000) coincide exactamente con el total de usuarios 
únicos de la plataforma, confirmando que el `JOIN` entre `users` y 
`user_activity` no perdió ni duplicó usuarios.

### 🔎 Hallazgo 1 — no existe cohorte de junio 2025

Se investigó inicialmente como una posible censura por la derecha (usuarios 
registrados muy cerca del fin del periodo de datos, sin tiempo suficiente 
para completar el chequeo de 21 días). Sin embargo, se confirmó con una 
consulta directa que **no existen registros de usuarios en junio** 
(`fecha_registro` de `users` va de 2025-01-01 a 2025-05-31). Esto contrasta 
con `orders` y `marketing`, que sí tienen actividad hasta el 30 de junio — 
sugiere que la adquisición de nuevos usuarios se detuvo un mes antes que el 
resto de la actividad de la plataforma, posiblemente por el cierre de una 
campaña de adquisición específica. Se recomienda validar esta hipótesis con 
el equipo de marketing.

### 🔎 Hallazgo 2 — la retención se mantiene estable, sin caída progresiva clara

A diferencia de un patrón típico de caída semana a semana, la retención se 
mantiene relativamente estable entre 40% y 44% en las 3 semanas medidas, en 
las 5 cohortes. No hay una tendencia consistente de mejora ni deterioro mes 
a mes. Esto sugiere un núcleo de usuarios con comportamiento estable, más que 
una fuga acelerada de usuarios después del registro.

### Nota sobre alcance de los datos

La tabla `user_activity` contiene un cuarto punto de chequeo a los 28 días 
(semana 4) que no se utilizó en este análisis, por estar fuera del alcance 
definido en el objetivo (semanas 1–3). Queda disponible para extender este 
análisis en el futuro si el negocio lo requiere.

---

## 🔹 Paso 5: Validar si los cambios generan impacto (test estadístico)

🎯 **Objetivo:** Evaluar si la modificación en la UI del checkout impacta la **tasa de conversión de compra**.

---

1. **Analizar el dataset** `experiment_checkout_ui.csv` para identificar la métrica principal **conversion**.
   - La métrica **conversion** es 1 si el usuario completó la compra, 0 si no.    
2. **Plantear la hipótesis estadística**     
3. **Aplicar el test estadístico adecuado** 
4. **Interpretar el resultado**  

---
Hipótesis estadística
   - **H₀ (Hipótesis nula):** ...
   - **H₁ (Hipótesis alternativa):** ...
   
**Test estadístico:** ...  
**Nivel de significancia alpha:** ...

In [84]:
import pandas as pd

url_experiment = 'https://practicum-content.s3.amazonaws.com/datasets/experiment_checkout_ui.csv'
experiment = pd.read_csv(url_experiment)

print(experiment.shape)
print(experiment.info())
experiment.head()


(10000, 7)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id_usuario       10000 non-null  object 
 1   variante         10000 non-null  object 
 2   convirtio        10000 non-null  int64  
 3   dispositivo      10000 non-null  object 
 4   pais             10000 non-null  object 
 5   duracion_sesion  10000 non-null  float64
 6   timestamp        10000 non-null  object 
dtypes: float64(1), int64(1), object(5)
memory usage: 547.0+ KB
None


,id_usuario,variante,convirtio,dispositivo,pais,duracion_sesion,timestamp
0,exp_user_0,tratamiento,0,mobile,Argentina,114.41,2025-03-28
1,exp_user_1,tratamiento,0,desktop,Mexico,170.03,2025-01-15
2,exp_user_2,control,1,mobile,Colombia,140.21,2025-03-18
3,exp_user_3,tratamiento,0,mobile,Colombia,151.45,2025-06-03
4,exp_user_4,tratamiento,0,desktop,Mexico,299.96,2025-01-12


### 1 paso: verificar el tamaño de cada grupo ###

In [85]:
print(experiment['variante'].value_counts())

tratamiento    5035
control        4965
Name: variante, dtype: int64


### 2 paso: Calcular la tasa de conversión por grupo

In [86]:
tasas_conversion = experiment.groupby('variante')['convirtio'].agg(['sum', 'count', 'mean'])
tasas_conversion.columns = ['conversiones', 'total_usuarios', 'tasa_conversion']
print(tasas_conversion)

             conversiones  total_usuarios  tasa_conversion
variante                                                  
control               779            4965         0.156898
tratamiento           820            5035         0.162860


### 3 paso: El test estadístico — prueba de proporciones (z-test) ###

In [87]:
from statsmodels.stats.proportion import proportions_ztest

# Extraemos los valores que necesita la función
conversiones = tasas_conversion['conversiones'].values
totales = tasas_conversion['total_usuarios'].values

z_stat, p_value = proportions_ztest(count=conversiones, nobs=totales)

print(f'Estadístico Z: {z_stat:.4f}')
print(f'P-valor: {p_value:.4f}')

Estadístico Z: -0.8133
P-valor: 0.4161


### 4 paso: comparar el p-valor contra alpha ###

In [88]:
alpha = 0.05

if p_value < alpha:
    print(f'p-valor ({p_value:.4f}) < alpha ({alpha}) → Se rechaza H0')
    print('Existe evidencia estadísticamente significativa de que el cambio de UI afecta la conversión.')
else:
    print(f'p-valor ({p_value:.4f}) >= alpha ({alpha}) → No se rechaza H0')
    print('No hay evidencia suficiente de que el cambio de UI afecte la conversión.')

p-valor (0.4161) >= alpha (0.05) → No se rechaza H0
No hay evidencia suficiente de que el cambio de UI afecte la conversión.


In [89]:
tasas_por_dispositivo = experiment.groupby(['dispositivo', 'variante'])['convirtio'].mean().unstack()
tasas_por_dispositivo['diferencia'] = tasas_por_dispositivo['tratamiento'] - tasas_por_dispositivo['control']
print(tasas_por_dispositivo)

variante      control  tratamiento  diferencia
dispositivo                                   
desktop      0.183812     0.181926   -0.001886
mobile       0.129426     0.143543    0.014116


In [90]:
from statsmodels.stats.proportion import proportions_ztest

for dispositivo in experiment['dispositivo'].unique():
    subset = experiment[experiment['dispositivo'] == dispositivo]
    stats = subset.groupby('variante')['convirtio'].agg(['sum', 'count'])
    
    z_stat, p_val = proportions_ztest(count=stats['sum'].values, nobs=stats['count'].values)
    
    print(f'\n--- {dispositivo} ---')
    print(stats)
    print(f'P-valor: {p_val:.4f}')


--- mobile ---
             sum  count
variante               
control      318   2457
tratamiento  359   2501
P-valor: 0.1478

--- desktop ---
             sum  count
variante               
control      461   2508
tratamiento  461   2534
P-valor: 0.8625


## 📋 Resumen — Paso 5: Test estadístico (experimento de UI en checkout)

**Dataset:** `experiment_checkout_ui.csv` — 10,000 usuarios, sin valores nulos
**Grupos:** control (4,965 usuarios) vs. tratamiento (5,035 usuarios) — balanceados correctamente

---

### Hipótesis estadística

- **H₀ (nula):** no existe diferencia real en la tasa de conversión entre el diseño 
  actual del checkout (control) y el nuevo diseño (tratamiento).
- **H₁ (alternativa):** existe una diferencia real en la tasa de conversión entre 
  ambos diseños.

**Test aplicado:** prueba Z de dos proporciones (`proportions_ztest`), adecuada 
para comparar tasas de conversión (variable binaria) entre 2 grupos independientes.
**Nivel de significancia:** alpha = 0.05

---

### Resultado principal

| Grupo | Conversiones | Usuarios | Tasa de conversión |
|---|---|---|---|
| Control | 779 | 4,965 | 15.69% |
| Tratamiento | 820 | 5,035 | 16.29% |

**Estadístico Z:** -0.8133
**P-valor:** 0.4161

Como el p-valor (0.4161) es mayor que alpha (0.05), **no se rechaza H₀**.

**Conclusión:** con los datos disponibles, no hay evidencia estadísticamente 
significativa de que el nuevo diseño de checkout tenga un efecto sobre la tasa 
de conversión. La diferencia observada (+0.6 puntos porcentuales a favor del 
tratamiento) es consistente con variabilidad aleatoria de la muestra, y no 
permite afirmar que el cambio de UI mejora ni empeora la conversión.

*Nota de interpretación: este resultado no demuestra que ambos diseños sean 
"iguales" — solo indica que, con el tamaño de muestra actual (10,000 usuarios), 
no se detectó evidencia suficiente de una diferencia real.*

---

### Análisis exploratorio por dispositivo (no confirmatorio)

| Dispositivo | Control | Tratamiento | Diferencia | P-valor |
|---|---|---|---|---|
| Mobile | 12.94% | 14.35% | +1.41 pp | 0.1478 |
| Desktop | 18.38% | 18.19% | -0.19 pp | 0.8625 |

Aunque el subgrupo de mobile mostraba visualmente la mayor diferencia entre 
variantes, el test formal confirma que **tampoco es estadísticamente 
significativo** — ni siquiera al nivel de significancia estándar. Este 
desglose se reporta como exploratorio, no como evidencia confirmatoria: 
dividir el experimento en subgrupos reduce el tamaño de muestra de cada uno, 
lo que dificulta detectar diferencias reales y aumenta el riesgo de interpretar 
ruido aleatorio como una señal genuina (problema de comparaciones múltiples).

---

### Recomendación de negocio

No se encontró evidencia suficiente para justificar el lanzamiento del nuevo 
diseño de checkout basándose en este experimento. Antes de descartarlo por 
completo, se recomienda:
1. Evaluar si el experimento tuvo suficiente **poder estadístico** para 
   detectar una diferencia del tamaño observado (0.6 puntos porcentuales) — 
   es posible que se necesite una muestra considerablemente mayor.
2. Si el negocio tiene interés específico en el segmento mobile (donde se 
   observó la mayor diferencia, aunque no significativa), diseñar un 
   experimento dedicado a ese segmento, con un tamaño de muestra calculado 
   específicamente para detectar ese efecto.

---

## 🔹 Paso 6: Comunicar los resultados (Dashboard en BI)

🎯 **Objetivo**:  
Crear un dashboard que muestre de manera clara y visual los resultados del análisis de ventas, costos, marketing y conversión. 

Se usarán los CSVs limpios del Paso 1:

- `orders_clean.csv`  
- `catalog_clean.csv`  
- `marketing_clean.csv`

---

1️⃣ Preparación de los datos
1. Cargar los CSVs en Power BI o Tableau.
2. Revisar relaciones:
   - `orders.nombre_producto` → `catalog.nombre_producto`
   - `orders.fecha_pedido` → tabla de fechas (crear calendario para análisis temporal)
   - `orders.fecha_pedido` → `dim_fecha.date`
3. Crear columnas calculadas necesarias
4. Crear tabla de fechas para poder calcular comparaciones YTD, YoY o períodos anteriores (`Previous Year`, `Previous Month`).

---

2️⃣ Dashboard 1: Overview Ejecutivo
**KPIs principales a mostrar:**
- Revenue total
- Profit total
- Gasto total en marketing
- Ticket promedio
- Cantidad promedio de productos por orden

**Visualizaciones sugeridas:**
- Tarjetas KPI para revenue, profit y gasto marketing
- Gráfico de líneas: evolución mensual de revenue o profit
- Gráfico de líneas YTD
- Gráfico de barras: revenue y profit por producto o categoría

---

 3️⃣ Dashboard 2: Detalle / Drill-through  
**Objetivo:** Permitir explorar los datos desde el KPI general hasta cada orden o producto.

**Visualizaciones sugeridas:**
- Tabla detallada de órdenes con:
  - producto, cantidad, revenue, cost, profit
  - color condicional (profit negativo en rojo, positivo en verde)
- Gráfico de barras por producto con medida `cantidad vendida`
- Drill-through: seleccionar un producto y ver todos los pedidos relacionados
- Filtros por fecha, categoría de producto, etc

---

## 🚀 Entrega Final

Comparte el acceso a tu Dashboard para revisión.   
Puedes entregar el Dashboard utilizando **Power BI o Tableau**.

Incluye **uno de los siguientes**:

- 🔗 Link público del dashboard publicado en **Power BI Service o Tableau Public / Tableau Cloud**
- 🔗 Link de **Google Drive o OneDrive** con el archivo del proyecto (`.pbix`) y los 3 csvs limpios.


### 📎 Enlace del Dashboard

In [ ]:

https://app.powerbi.com/view?r=eyJrIjoiYmQ1MWMzZTQtMWFlOC00MjZmLWIwMDEtODlkNzJiMDY5MmI4IiwidCI6IjI2NGEzYzA5LWUzOGQtNDdlZi04NzE0LWY3YzEwZGZiOThjZSIsImMiOjR9